[WIP] plasma5: 5.17.5 -> 5.18.5

petabyteboy opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

###### Todo list:

- [x] port patches / fix broken behaviour in these packages
  - [x] plasma-workspace
    - [x] 0001-startkde.patch
    - [x] 0002-absolute-wallpaper-install-dir.patch
  - [x] kde-gtk-config
    - [ ] follow-symlinks.patch
    - [x] gsettings.patch
  - [x] sddm plasma theme
    - [x] fix background image

- [x] fix broken packages:
  - [x] deepin.dde-kwin  
  - [x] deepin.startdde
  - [x] kdecoration-viewer


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
plasma5: 5.17.5 -> 5.18.5

cole-h commented on pkgs/development/libraries/qt-5/5.14/default.nix:
```diff
@@ -0,0 +1,174 @@
+    # Community port of the now unmaintained upstream qtwebkit.
+    qtwebkit = {
+      src = fetchFromGitHub {
+        owner = "annulen";
```
Just want to make sure you don't forget: https://github.com/NixOS/nixpkgs/pull/84232#discussion_r403618251

petabyteboy replied:
Yes thanks, I haven't marked it as resolved yet so I won't forget :)

petabyteboy replied:
This is not relevant to this PR anymore since I decided to keep the Qt update out of it.

petabyteboy commented:
nix-review output:

```
https://github.com/NixOS/nixpkgs/pull/84542
4 package marked as broken and skipped:
gazeboSimulator.sdformat gazeboSimulator.sdformat3 kst linuxPackages_hardkernel_4_14.systemtap

3 package failed to build:
deepin.dde-kwin deepin.startdde kdecoration-viewer

54 package built:
bluedevil breeze-grub breeze-gtk breeze-plymouth breeze-qt5 digikam discover kactivitymanagerd kde-cli-tools kde-gtk-config kde2-decoration kdeApplications.kdenlive kdeApplications.spectacle kdeplasma-addons kdev-php kdev-python kdevelop kdevelop-unwrapped kgamma5 kinfocenter kmenuedit krohnkite kscreen kscreenlocker ksshaskpass ksysguard kwallet-pam kwayland-integration kwin kwin-tiling kwrited plasma5.kdecoration plasma5.khotkeys plasma5.libkscreen plasma5.libksysguard lxqt.lxqt-config milou oxygen plasma-applet-volumewin7mixer plasma-browser-integration plasma-desktop plasma-integration plasma-nm plasma-pa plasma-vault plasma-workspace plasma-workspace-wallpapers plasma5.polkit-kde-agent plasma5.powerdevil plasma5.sddm-kcm plasma5.systemsettings plasma5.user-manager plasma5.xdg-desktop-portal-kde wacomtablet

```

petabyteboy commented on pkgs/desktops/plasma-5/plasma-workspace/0001-startkde.patch:
```diff
@@ -48,22 +48,7 @@ index e0f7004..8ac41fd 100644
  }
  
  QStringList allServices(const QLatin1String& prefix)
-@@ -184,14 +184,6 @@ void runEnvironmentScripts()
```
TODO: what's up with this

petabyteboy replied:
This is not required anymore since https://github.com/KDE/plasma-workspace/commit/3c1a34a26dbdf7c123782746881d914cd47b27b6

petabyteboy commented:
It actually works surprisingly well. The builtin wallpapers are now detected again and the system settings seem fine too.

![image](https://user-images.githubusercontent.com/3250809/78900951-e5018380-7a77-11ea-8a94-5f392cde4772.png)

peterhoeg commented:
Is this getting good to go by chance?

petabyteboy commented:
The gtk settings and deepin packages are broken (due to kwin update). I have spent some time on both but didn't find any easy fixes. If you want to bring this closer to being merged, feel free to take a look at those problems. I can not tell how much time I will be able to spend on this in the next days.

romildo commented:
> The gtk settings and deepin packages are broken (due to kwin update).

The deepin packages have stagnated on upstream. For almost an year they have not been updated. Deepin Linux is working privately on a new version of their Linux distribution with substantial changes. For this reason the packaging of the Deepin Desktop Environment is dormant on nixpkgs as well, waiting new releases from upstream.

For now I think you can mark both `deepin.dde-kwin` and `deepin.startdde` (which depends on `deepin.dde-kwin`) as broken.

cc @worldofpeace

worldofpeace commented:
I tried fixing dde-kwin and I tried:
* patching in https://git.archlinux.org/svntogit/community.git/tree/trunk/kwin-5.17.patch?h=packages/deepin-kwin 
* https://github.com/linuxdeepin/dde-kwin/pull/106.patch

and getting those to apply, but the build still fails at obscure spots.

@romildo It should be fine to mark as broken because we don't have the deepin desktop module, so no one is going to really miss anything.

Commit:
deepin.dde-kwin, deepin.startdde: mark as broken

These will not compile anymore with the latest version of kwin and
upstream development has moved to a private repository.

petabyteboy commented:
I have marked the failing deepin packages as broken and updated the todo list above with my latest findings.

worldofpeace commented:
> The deepin packages have stagnated on upstream. For almost an year they have not been updated. Deepin Linux is working privately on a new version of their Linux distribution with substantial changes. For this reason the packaging of the Deepin Desktop Environment is dormant on nixpkgs as well, waiting new releases from upstream.

Does this even mean it's open source anymore? Any sources on this?

romildo commented:
> Does this even mean it's open source anymore? Any sources on this?

[Deepin Linux](https://www.deepin.org/en/) home page does not give much information on that. But the [forum](https://bbs.deepin.org/forum.php?mod=forumdisplay&fid=70) has some clues. Unfortunately most of the messages are in chinese.

Recently they published a [beta](https://www.deepin.org/en/2020/04/15/deepin-20-beta/) of the next version (Deepin 20 beta), where they wrote:

> deepin is an open source desktop system developed for common users. You are free to download, distribute, modify and use it.

In a [post in the forum](https://bbs.deepin.org/forum.php?mod=redirect&goto=findpost&ptid=192041&pid=688385) they say:

> All the code had been published in the beta repo
but not pushed to github yet
They will push them when V20 official release

The company behind the Deepin Linux is also producing a commercial distribution (I do not remember its name now :) and they gave priority to it. That is the reason for the delay in the development of Deepin Linux. Those comments can be found in the forum if one looks for them.

DIzFer commented on pkgs/desktops/plasma-5/plasma-workspace/0002-absolute-wallpaper-install-dir.patch:
```diff
@@ -8,15 +8,16 @@ Subject: [PATCH 2/2] absolute-wallpaper-install-dir
  fontSize=10
--background=${CMAKE_INSTALL_PREFIX}/${WALLPAPER_INSTALL_DIR}/Next/contents/images/5120x2880.png
+-background=${KDE_INSTALL_FULL_WALLPAPERDIR}/Next/contents/images/5120x2880.jpg
 +background=${NIXPKGS_BREEZE_WALLPAPERS}/Next/contents/images/5120x2880.png
```
Hey, there's a whoopsie here. Notice how the upstream format changed to jpg.

petabyteboy commented:
Rebased, bumped to 5.18.5, fixed sddm wallpaper (thx @DIzFer!)

5.19.x is already out, but 5.18 is LTS and 5.19 requires Qt >=5.14.

Commit:
kdecoration-viewer: remove

Fails to build with Plasma >= 5.18 and there have not been any changes
in upstream since a long time.
According to the maintainer (@gnidorah) it can be removed.

petabyteboy commented:
I have marked kdecoration-viewer as broken for now, as it does not seem like an essential application and there is no fix available. (CC @gnidorah )

If someone needs it, it can be obtained from the release branch like this:

```
nix build -f https://github.com/nixos/nixpkgs/archive/nixos-20.03.tar.gz kdecoration-viewer
./result/bin/kdecoration-viewer
```

petabyteboy commented:
I could use some help with the gtk config page.
The plasma kde-gtk-config module was mostly rewritten from scratch. The menu was moved to the other Application Style code, and kde-gtk-config now only contains a kded plugin. If the settings menu detects that the kded kde-gtk-config module is loaded, it will show the settings page.
I don't know how to figure out if the kded module has been loaded or where exactly the problem is, but the settings page is not shown.

petabyteboy commented:
> @petabyteboy You're free to even remove the `kdecoration-viewer` entirely

Thanks for the feedback, I removed it.

worldofpeace commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20473,8 +20473,6 @@ in
 
   kdeconnect = libsForQt5.callPackage ../applications/misc/kdeconnect { };
 
-  kdecoration-viewer = libsForQt5.callPackage ../tools/misc/kdecoration-viewer { };
```
please add a `throw` in aliases for deprecation

petabyteboy replied:
Thanks! I forgot to commit that line.

cole-h commented on pkgs/desktops/plasma-5/plasma-desktop/default.nix:
```diff
@@ -1,3 +1,4 @@
+
```
Kinda weird, but I won't judge.

petabyteboy replied:
That was not on purpose :D
Thanks for catching it

cole-h reviewed (approved):
Diff LGTM (aside from a sneaky newline).

peterhoeg commented on pkgs/desktops/deepin/dde-kwin/default.nix:
```diff
@@ -133,6 +133,7 @@ mkDerivation rec {
   passthru.updateScript = deepin.updateScript { inherit pname version src; };
 
   meta = with stdenv.lib; {
+    broken = true;
```
Nitpick, but can we do `broken = lib.versionAtLeast kwindowsystem.version "5.18";` ? This makes it super clear why it's broken.

peterhoeg replied:
Or maybe `broken = lib.versionAtLeast (lib.versions.majorMinor kwindowsystem.version) "5.18";`

petabyteboy replied:
Both expressions don't work. If you find one that works I'll happily put it in.

petabyteboy replied:
`broken = versionAtLeast (versions.majorMinor (getVersion kwindowsystem)) "5.18";` seems to work

ttuegel commented on pkgs/desktops/plasma-5/kde-gtk-config/default.nix:
```diff
@@ -13,7 +13,6 @@ mkDerivation {
     ki18n kio glib gtk2 gtk3 karchive kcmutils kconfigwidgets kiconthemes
     knewstuff gsettings-desktop-schemas
   ];
-  patches = [ ./patches/follow-symlinks.patch ./patches/gsettings.patch ];
```
Have we confirmed that these packages are actually not necessary? I recall, in particular, that `kde-gtk-config` was completely broken without `follow-symlinks.patch`.

petabyteboy replied:
Please refer to this comment on the state of gtk-config:

> I could use some help with the gtk config page.
> The plasma kde-gtk-config module was mostly rewritten from scratch. The menu was moved to the other Application Style code, and kde-gtk-config now only contains a kded plugin. If the settings menu detects that the kded kde-gtk-config module is loaded, it will show the settings page.
> I don't know how to figure out if the kded module has been loaded or where exactly the problem is, but the settings page is not shown.

So no, without patches it doesn't work, but since source was 100% replaced with new code upstream, the patches won't apply and can not be easily rewritten.

petabyteboy replied:
The answer to your question is: No, we haven't confirmed. Quite the opposite, we have confirmed that it doesn't work without patches, which is also the reason why figuring out the new way to do this is part of the todo list in the initial comment.

ttuegel replied:
As I recall, the patches mainly affected the code for the settings menu. The code that needs to be patched has probably moved to `systemsettings`. I guess we may _also_ need to patch the `kded` module, but I can't imagine what that module is actually doing: setting the GTK theme is a matter of writing a few static files, so I don't know why we need a running daemon to do that.

petabyteboy replied:
Some more information on this new "Plasma GTKd" is available here:
https://phabricator.kde.org/D24743

petabyteboy replied:
And the new settings page is added here: https://phabricator.kde.org/D26583

petabyteboy replied:
I think the new code already follows symlinks, at least the only place where I can find something that looks similar to the code that was previously patched (plasma-desktop/kcms/style/gtkthemesmodel.cpp L. 154) uses QDir().entryList without passing QDir::NoSymlinks.

ttuegel commented:
I am investigating the GTK theme issue now. My first guess is that it is due to the missing dependency on `xsettingsd`.

ttuegel committed:
kde-gtk-config: add xsettingsd input

ttuegel commented:
> I am investigating the GTK theme issue now. My first guess is that it is due to the missing dependency on `xsettingsd`.

Actually, it has nothing to do with `kde-gtk-config` at all. It seems that `kded` is not running, or at least `systemsettings` cannot connect to it. I discovered this when I tried to configure `Background Services`.

ttuegel commented:
Apparently I already fixed this? https://github.com/NixOS/nixpkgs/pull/40447
But now the patch must move. :roll_eyes:

FRidh commented:
deepin was removed in https://github.com/NixOS/nixpkgs/pull/96610.

ttuegel committed:
kde-gtk-config: set GSettings schemas path

ttuegel commented:
I fixed `kde-gtk-config`, but I'm still tracking down some missing dependencies.

ttuegel committed:
kdeplasma-addons: add kpurpose and qtwebengine as inputs

ttuegel committed:
Merge branch 'master' into feature/plasma-5-18

ttuegel committed:
kinfocenter: add kirigami2 input

ttuegel committed:
ksysguard: add libcap and libpcap inputs

ttuegel committed:
kwin: add qtquickcontrols2 and qtvirtualkeyboard inputs

ttuegel committed:
plasma-desktop: add qqc2-desktop-style input

ttuegel committed:
kquickcharts: init at 5.73.0

ttuegel committed:
plasma-workspace: add kquickcharts and appstream-qt inputs

ttuegel committed:
plasma-pa: add sound-theme-freedesktop input

ttuegel committed:
xdg-desktop-portal-kde: add epoxy and mesa inputs

ttuegel committed:
powerdevil: update patches

ttuegel commented:
We are still affected by #96055, but I don't see that as a reason to block merging as soon as the check pass again!

ttuegel commented:
Thanks everyone who helped this happen, especially @petabyteboy ! :tada:
